### PR TITLE
Add base map file format

### PR DESCRIPTION
**Pull request by ihhub**
This is a base map file format what will be modified in the future many times. However, this is a step stone towards a full editor.

To test map format saving you can add this code:
```cpp
StreamFile file;
file.open( "test.map", "wb" );
Maps::Map_Format::MapFormat map;
Maps::saveMapInEditor( map );
file << map;
file.close();
```

To test map format loading you can add this code:
```cpp
StreamFile file;
file.open( "test.map", "rb" );
Maps::Map_Format::MapFormat map;
file >> map;
Maps::readMapInEditor( map );
```

### CODE REVIEW

**Comment by ihhub** [comment]
> I think, it`s better to use "h2m" filename extenstion for expanded fheroes2 maps, just in line for other HoMM games ("h3m" for Heroes 3, "h4m" for Heroes 4, "h5m" for Heroes 5 etc,).

This pull request doesn't add any new map file extension.

File: src/fheroes2/maps/map_object_info.h [hunk 1]
```diff
@@ -117,6 +117,10 @@ namespace Maps
     // These groups do not correlate with the original Editor.
     enum class ObjectGroup : int32_t
     {
+        // These groups are not being used by the Editor directly but they are still a part of a tile.
+        Roads,
+        Streams,
+
         // Landscape objects.
         Landscape_Mountains,
```
Comment (oleg-derevenetz) [inline src/fheroes2/maps/map_object_info.h:125]: It doesn't relate directly to this PR, just a question while I'm taking a look on it. We usually apply either `UPPERCASE_UPPERCASE` naming scheme for enum constants (in the older code) or `PascalCase` (in a newer code). Why the new style? :)

Comment (ihhub) [reply src/fheroes2/maps/map_object_info.h:125]: No idea. I think I should change it :)

Comment (ihhub) [reply src/fheroes2/maps/map_object_info.h]: I addressed this.

**Review by oleg-derevenetz — APPROVED** [review]
Looks good to me, at least as an initial format for further development.

File: src/fheroes2/world/world.cpp [hunk 2]
```diff
@@ -441,22 +431,12 @@ void World::NewMaps( int32_t sw, int32_t sh )
...
-        mp2tile.level1ObjectUID = 0; // means that there's no object on this tile.
-        mp2tile.level2ObjectUID = 0;
-
-        vec_tiles[i].Init( static_cast<int32_t>( i ), mp2tile );
+        vec_tiles[i] = {};
+
+        vec_tiles[i].setIndex( static_cast<int32_t>( i ) );
+
+        const uint8_t terrainFlag = static_cast<uint8_t>( Rand::Get( 0, 3 ) );
+        vec_tiles[i].setTerrain( static_cast<uint16_t>( Rand::Get( 16, 19 ) ), terrainFlag & 1, terrainFlag & 2 );
```
[the remark this answers is not in the document]
Comment (ihhub) [reply src/fheroes2/world/world.cpp:439]: Hi @Districh-ru , I tried this approach and noticed groups of repeated patterns on water. I think it is better to leave it as it is to have more or less uniformed water surface.

[the remark this answers is not in the document]
Comment (Districh-ru) [reply src/fheroes2/world/world.cpp:439]: Ok, lets leave it for now.

**Pull request merged**